Add song select beatmap leaderboard display

frenzibyte opened the pull request:
- [x] Depends on https://github.com/ppy/osu/pull/32840
- [x] Depends on https://github.com/ppy/osu/pull/32913 (and application of refactors as per https://github.com/ppy/osu/pull/32844#discussion_r2053689750)
- Split from https://github.com/ppy/osu/pull/32715

Preview:

https://github.com/user-attachments/assets/b6d7956f-cdd9-4aa0-a945-8247e8a7b903

Notes:
 - @peppy Seeing how [`BeatmapLeaderboard`](https://github.com/ppy/osu/blob/bb30d86401eea66632942b5c5dac3364bd47369b/osu.Game/Screens/Select/Leaderboards/BeatmapLeaderboard.cs#L136-L151) implements score fetching, I've attempted to follow suit and do the exact same to have full functionality support in https://github.com/ppy/osu/pull/32844/commits/de9146c55efcfa24c42cfe7675b885d7dd4ae3ca. Feel free to revert / force-push this commit away if you still think it shouldn't be done here.
 - If a score card goes partially off-screen, then resizing the width back and fourth doesn't work correctly. This is happening due to the parent of `rankLabelStandalone` getting masked away completely, and therefore not processing transforms to that component. I have no idea how to fix this one because the parent here is a cell container of `GridContainer`, somehow `GridContainer.CellContainer` needs to have `IsMaskedAway` never set to false.

   https://github.com/user-attachments/assets/84b3cdf2-1bde-4c91-a230-3987a9ca31fa

 - Total leaderboard scores count that was added in https://github.com/ppy/osu-web/pull/12018 has not been implemented yet, as it requires minor changes to `LeaderboardManager` to carry the value across to consumers. To be done in a follow-up.

frenzibyte committed:
Update beatmap leaderboard score design to match new metrics

frenzibyte committed:
Add new beatmap leaderboard score tooltip

frenzibyte committed:
Add beatmap leaderboard wedge

frenzibyte committed:
Implement score fetch functionality

Copied logic from `BeatmapLeaderboard`.

peppy committed:
Merge branch 'master' into song-select-v2-wedges-leaderboard

peppy committed:
Fix regression from copy waste

peppy commented:
> @peppy Seeing how [BeatmapLeaderboard](https://github.com/ppy/osu/blob/bb30d86401eea66632942b5c5dac3364bd47369b/osu.Game/Screens/Select/Leaderboards/BeatmapLeaderboard.cs#L136-L151) implements score fetching, I've attempted to follow suit and do the exact same to have full functionality support in https://github.com/ppy/osu/commit/de9146c55efcfa24c42cfe7675b885d7dd4ae3ca. Feel free to revert / force-push this commit away if you still think it shouldn't be done here.

You copied a bug and missed a test, see https://github.com/ppy/osu/pull/32816. Fixed via https://github.com/peppy/osu/commit/7a18a771b3e557e2f64a604636de2f70f99d4952.

peppy committed:
Seal `BeatmapLeaderboardScore` for now

We'll need to figure out what to do in multiplayer cases in the future,
but the hope is that it can be done without further subclassing if
possible.

peppy committed:
Fix non-sheared test showing sheared drawables

peppy commented:
Cases like https://github.com/ppy/osu/pull/32844/commits/26f2703688258f010f31699dd2052584b67a2225 make me very wary that you haven't run through all added tests on a visual level.

bdach commented on osu.Game/Screens/SelectV2/BeatmapLeaderboardWedge.cs:
```diff
@@ -0,0 +1,390 @@
+            SetState(LeaderboardState.Retrieving);
+
+            if (beatmap.IsDefault)
+            {
+                SetState(LeaderboardState.NoneSelected);
+                return;
+            }
+
+            var fetchBeatmapInfo = beatmap.Value.BeatmapInfo;
+            var fetchRuleset = ruleset.Value ?? fetchBeatmapInfo.Ruleset;
+
+            if (!api.IsLoggedIn && isOnlineScope)
+            {
+                SetState(LeaderboardState.NotLoggedIn);
+                return;
+            }
+
+            if (!fetchRuleset.IsLegacyRuleset())
+            {
+                SetState(LeaderboardState.RulesetUnavailable);
+                return;
+            }
+
+            if ((fetchBeatmapInfo.OnlineID <= 0 || fetchBeatmapInfo.Status <= BeatmapOnlineStatus.Pending) && isOnlineScope)
+            {
+                SetState(LeaderboardState.BeatmapUnavailable);
+                return;
+            }
+
+            if (Scope.Value.RequiresSupporter(FilterBySelectedMods.Value) && !api.LocalUser.Value.IsSupporter)
+            {
+                SetState(LeaderboardState.NotSupporter);
+                return;
+            }
+
+            if (Scope.Value == BeatmapLeaderboardScope.Team && api.LocalUser.Value.Team == null)
+            {
+                SetState(LeaderboardState.NoTeam);
+                return;
+            }
```
All of this should be gone and `LeaderboardManager` should be relied on to produce these error conditions instead. See https://github.com/ppy/osu/pull/32863

peppy replied:
I'm going to PR a minor refactor to the base class before we copy it here, because I rewrote some of the logic in frenzi's PR sequence around this code without realising it still exists in the old implementation.

peppy replied:
Revisit this after https://github.com/ppy/osu/pull/32913

peppy replied:
I've updated the logic @bdach a second set of eyes on it would be appreciated.

bdach replied:
https://github.com/ppy/osu/pull/32844/commits/3b2e8281b4e3f2ecdf5d6821427973d9004bb01d aside, looks good.

peppy committed:
Make leaderbaord animation barely bareable

peppy committed:
Don't animate on initial mode display

This removes a lot of movement, but honestly it didn't feel good in the
first place. If anything I'll come back with a second-pass animation
pass on this.

frenzibyte commented:
> Cases like https://github.com/ppy/osu/pull/32844/commits/26f2703688258f010f31699dd2052584b67a2225 make me very wary that you haven't run through all added tests on a visual level.

That's a test from master branch and doesn't look like one I need to carefully pay attention to, it displays non-sheared version while I'm working completely on sheared version. I'm not sure I have to be responsible for that.

peppy committed:
Merge branch 'master' into song-select-v2-wedges-leaderboard

peppy committed:
Update fetch logic to match existing leaderboard

Also handles some display edge cases where scores may overlap
placeholder or do other weird things.

peppy commented:
@frenzibyte we have four mod tooltips now. the one in the beatmap score is only used there:

![JetBrains Rider 2025-04-23 at 09 56 36](https://github.com/user-attachments/assets/a417c232-fe99-49d0-b9c5-a3fc92b090b4)

what's the forward direction here?

peppy committed:
Begin to fix eyesore code in `BeatmapLeaderboardScore`

peppy commented on osu.Game/Screens/SelectV2/BeatmapLeaderboardScore.cs:
```diff
@@ -697,52 +746,62 @@ private void load(OsuColour colours, OverlayColourProvider colourProvider)
         }
 
-        private sealed partial class MoreModSwitchTiny : CompositeDrawable
+        private sealed partial class MoreModSwitchTiny : CompositeDrawable, IHasPopover
```
A button called a switch, with no sounds attached.

frenzibyte replied:
Has been removed as part of moving from tiny mod icons to the old full ones.

peppy commented:
`BeatmapLeaderboardScore` is, nothing short of a pile of turd.

Yes, I get that it's not your original code, but it also feels like you've just turned a blind eye to it while also refactoring a good portion of the class.

As one example, it has a heap of local animations which are only activated in the `TestSceneBeatmapLeaderboardScore`, but never seen outside of this. I'd have a hope that when you bring classes like this online – even in a design pass – you'd catch on to these things and amend them.

peppy committed:
Remove all animations from `BeatmapLeaderboardScore` and fix more eyesore code

bdach committed:
Remove double binding

bdach commented on osu.Game/Screens/SelectV2/BeatmapLeaderboardWedge.cs:
```diff
@@ -0,0 +1,374 @@
+                        {
+                            RelativeSizeAxes = Axes.X,
+                            AutoSizeAxes = Axes.Y,
+                            Margin = new MarginPadding { Top = 4f, Bottom = 180f },
```
Don't want to get too involved in the design side of things, but what is this bottom margin spec doing? I couldn't help but notice when checking the test scene that the margin makes it so that when you scroll all the way down there's a huge gap:

![Screenshot 2025-04-23 at 13 23 06](https://github.com/user-attachments/assets/146af139-e650-4c6f-addb-75df7090ed3d)

Seems much too weird for me to not say anything?

frenzibyte replied:
It's enough bottom bleed to allow the last card to show fully without being cut off from the left side.

peppy replied:
Should probably be commented as such!

frenzibyte replied:
Has been resolved in https://github.com/ppy/osu/pull/32844/commits/bd58aac9cca8b410b550ce6fa8afcdfb68a14b5d

peppy committed:
Begin to fix `BeatmapLeaderboardWedge` code quality

peppy committed:
Update 404ing cover image URLs

peppy committed:
Fix user covers not loading if one corner is off-screen

peppy committed:
Merge branch 'master' into song-select-v2-wedges-leaderboard

peppy commented:
would it not make more sense to have the score popover show all mods with all details rather than having a local clickable expand thing just for mods for some arbitrary reason?

peppy committed:
Add tests with mods with adjusted settings

frenzibyte committed:
Bring back full mod icons

frenzibyte committed:
Add DA mod with custom adjustment in new score test scene

frenzibyte committed:
Adjust right-side content layout to mask mods

peppy committed:
Remove colour lightening of judgement colours

I'm not sure why this is a thing but let's not do it without proper
rationale.

peppy reviewed (approved):
50% confidence
